File: epi_judge_java/epi/LowestCommonAncestorWithParent.java
package epi;
import epi.test_framework.BinaryTreeUtils;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import epi.test_framework.TestFailure;
import epi.test_framework.TimedExecutor;
public class LowestCommonAncestorWithParent {

  public static BinaryTree<Integer> lca(BinaryTree<Integer> node0,
                                        BinaryTree<Integer> node1) {
	  // Implement this placeholder.
	  // alog:
	  // 1. calc the height of node0 and node1.
	  // 2. move from the node with the greater height up till it reaches the level of the other node.
	  // 3. for simultaneously in the two nodes up, until the parent of the two nodes are equal. this is the LCA
	  // edge cases
	  // if node0 == node1 -> the LCA is one of them.

	  if (node0 == null || node1 == null) {
	  	return null;
	  }
	  int h0 = calcHeight(node0);
	  int h1 = calcHeight(node1);
	  int heightDiff = Math.abs(h1 - h0);
	  // make sure h1 >= h0 and node1 to have h1 and node 0 to have h0
	  if (h1 < h0) {
		  BinaryTree<Integer> tempNode = node1;
		  node1 = node0;
		  node0 = tempNode;
	  }

	  // go up for nodeA till it has the same height as NodeB;
	  while (heightDiff-- > 0) { // todo: can we use h1-- ?
		  node1 = node1.parent;
	  }

	  // here node0 and node1 have the same height;
	  while (node0 != node1) {
		  node0 = node0.parent;
		  node1 = node1.parent;
	  }
	  return node0;
  }

	private static int calcHeight(BinaryTree<Integer> node) {
		// use part to go up the tree till the parent is null;
		if (node == null) {
			return 0;
		}
		int ret = 0;
		while (node != null) {
			node = node.parent;
			ret++;
		}
		return ret;
	}
  @EpiTest(testDataFile = "lowest_common_ancestor.tsv")
  public static int lcaWrapper(TimedExecutor executor, BinaryTree<Integer> tree,
                               Integer key0, Integer key1) throws Exception {
    BinaryTree<Integer> node0 = BinaryTreeUtils.mustFindNode(tree, key0);
    BinaryTree<Integer> node1 = BinaryTreeUtils.mustFindNode(tree, key1);

    BinaryTree<Integer> result = executor.run(() -> lca(node0, node1));

    if (result == null) {
      throw new TestFailure("Result can not be null");
    }
    return result.data;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "LowestCommonAncestorWithParent.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_python/lowest_common_ancestor_with_parent.py
import functools
from typing import Optional

from binary_tree_with_parent_prototype import BinaryTreeNode
from test_framework import generic_test
from test_framework.binary_tree_utils import must_find_node
from test_framework.test_failure import TestFailure
from test_framework.test_utils import enable_executor_hook


def lca(node0: BinaryTreeNode,
        node1: BinaryTreeNode) -> Optional[BinaryTreeNode]:
    # TODO - you fill in here.
    # for the fist node
    # pupulate dict containing the nodes path in revese order till we reach the root
    seen = set()
    curr = node0
    while curr:
        seen.add(curr)
        curr = curr.parent

    # for the 2nd node, check if it is in the map
    curr = node1
    while curr:
        if curr in seen:
            return curr
        curr = curr.parent

    return None


@enable_executor_hook
def lca_wrapper(executor, tree, node0, node1):
    result = executor.run(
        functools.partial(lca, must_find_node(tree, node0),
                          must_find_node(tree, node1)))

    if result is None:
        raise TestFailure('Result can\'t be None')
    return result.data


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('lowest_common_ancestor_with_parent.py',
                                       'lowest_common_ancestor.tsv',
                                       lca_wrapper))
